wagtail_newsletter/actions.py

```diff
@@ -58,4 +70,15 @@ def send_campaign(request, page: NewsletterPageMixin) -> None:
 
     backend = campaign_backends.get_backend()
     backend.send_campaign(page.newsletter_campaign)
+
+    _log(page, "wagtail_newsletter.send_campaign")
+
     messages.success(request, "Newsletter campaign is now sending")
+
+
+def _log(instance, action, **kwargs):
+    user = get_active_log_context().user
+    uuid = get_active_log_context().uuid
+    return NewsletterLogEntry.objects.log_action(  # type: ignore
+        instance, action, page=instance, user=user, uuid=uuid, **kwargs
+    )
```

Comment:
I was wondering why we needed this instead of the standard `log()` function from Wagtail. I assume this is because Wagtail's `log()` uses the `LogActionRegistry` that maps a model to its corresponding `*LogEntry` model – which means passing a page instance would make it use `PageLogEntry` rather than `NewsletterLogEntry` we create in this PR.

This would likely lead to an issue: the `NewsletterLogEntry` isn't registered in the registry (see the last code snippet in https://docs.wagtail.org/en/stable/extending/audit_log.html#log-models), so any instances of the model won't show up in the site-wide history report. Or if you don't care about showing these entries in the Site history report, then I guess you can consider this a non-issue.

The registry, as it currently stands, is a one-to-one mapping from a model to the log entry model. We can't add `NewsletterLogEntry` to the registry, because we don't have a model to associate it with. If we use `Page`, it would overwrite the `Page` <-> `PageLogEntry` association. We also can't register it with the `NewsletterPageMixin` abstract model because we can't really "cast" the page instance to a `NewsletterPageMixin` instance (in such a way that `type(page)` returns `NewsletterPageMixin`) so the registry knows it should use `NewsletterLogEntry` instead of `PageLogEntry`.

As a result, our `NewsletterLogEntry` here is only a "partial" implementation of a custom log entry model.

I knew you discussed this with Dan, and I do agree that generally it is a good idea to use a custom log model if you're not mutating the page itself. However, if we look at how Wagtail's audit logs work in general, I think that only makes sense if you have _another_ model where the mutation takes place.

For example, if we had a model that represents a newsletter campaign for a page (e.g. the fields in `NewsletterPageMixin` as a separate model that has a `ForeignKey` to `Page`), we could map the `NewsletterLogEntry` to that model instead. But we don't have such model, which leads me to think maybe for the log entries we should have piggybacked off the `PageLogEntry` model instead. And since the actions are done in the page editor anyway, perhaps it's not such a stretch to log the newsletter actions directly against the page...

It would also mean you wouldn't have to set up a new log table, its utility functions, a separate history view etc. You've already done the work though... so I'm finding it hard to essentially tell you to undo all this 😅 

Alternatively I suppose we can use `NewsletterLogEntry` as-is, be satisfied that it doesn't show up on Site history report, and (maybe) eventually fix by adding the log entry model to the registry if and when we have a suitable model for it.

What do you think we should do here?

Reply:
> The registry, as it currently stands, is a one-to-one mapping from a model to the log entry model. We can't add `NewsletterLogEntry` to the registry, because we don't have a model to associate it with. If we use `Page`, it would overwrite the `Page` <-> `PageLogEntry` association. We also can't register it with the `NewsletterPageMixin` abstract model because we can't really "cast" the page instance to a `NewsletterPageMixin` instance (in such a way that `type(page)` returns `NewsletterPageMixin`) so the registry knows it should use `NewsletterLogEntry` instead of `PageLogEntry`.

Yep, that's precisely the reason why I'm using a separate logging code path. I didn't realise that log entries are shown in a site-wide history report though, and I think it's problematic to exclude newsletter logs from there.

> For example, if we had a model that represents a newsletter campaign for a page (e.g. the fields in `NewsletterPageMixin` as a separate model that has a `ForeignKey` to `Page`), we could map the `NewsletterLogEntry` to that model instead.

That is true. I'm hesitant to set up a model just so that we can log actions separately. On the other hand, it might make [the complexity of this revision gymnastics](https://github.com/wagtail/wagtail-newsletter/blob/30a64d97ec93803d4dfa249ea8e703f6af425131/wagtail_newsletter/models.py#L126-L145) go away. I'm tempted to have a go at this.

> But we don't have such model, which leads me to think maybe for the log entries we should have piggybacked off the `PageLogEntry` model instead. And since the actions are done in the page editor anyway, perhaps it's not such a stretch to log the newsletter actions directly against the page...
> 
> It would also mean you wouldn't have to set up a new log table, its utility functions, a separate history view etc. You've already done the work though... so I'm finding it hard to essentially tell you to undo all this 😅

To be honest, I'm happy to save entries to `PageLogEntry` (or `ModelLogEntry`, if that works out). The code is less complicated, therefore easier to maintain. Given the level of hackery involved in getting `NewsletterHistoryView` to work across 4 versions of Wagtail, I'd rather not have it in the codebase 😅

Reply:
> That is true. I'm hesitant to set up a model just so that we can log actions separately. On the other hand, it might make [the complexity of this revision gymnastics](https://github.com/wagtail/wagtail-newsletter/blob/30a64d97ec93803d4dfa249ea8e703f6af425131/wagtail_newsletter/models.py#L126-L145) go away. I'm tempted to have a go at this.

Yep, that's out of scope for this PR I think. You're welcome to experiment separately though 😄

Reply:
> In that case I'd say let's try that... 😃 Not sure if `ModelLogEntry` would work though, as you'd have the same problems unless you have a separate model to log against.

Yeah, that was assuming I'd create a new model, but you're right, it's out of scope. For now I've changed the code to log straight to `PageLogEntry`.